### PR TITLE
Measurement page

### PR DESCRIPTION
**Pull request by msdrey**

### CODE REVIEW

File: src/canvas/Document.tsx [hunk 1]
```diff
@@ -2,9 +2,11 @@ import { PatternPath } from './PatternPath';
 
 export class Document implements IDocument {
     private _patternPaths: PatternPath[];
+    private _sizeRatio: undefined | number; // in pixels per inch
```
Comment (SarahPappas) [inline src/canvas/Document.tsx]: please use null instead of undefined

File: src/canvas/Document.tsx [hunk 2]
```diff
@@ -2,9 +2,11 @@ import { PatternPath } from './PatternPath';
...
 
     constructor () {
         this._patternPaths = new Array<PatternPath>();
+        this._sizeRatio = undefined;
```
Comment (SarahPappas) [inline src/canvas/Document.tsx]: Please use null

File: src/canvas/Document.tsx [hunk 3]
```diff
@@ -28,4 +30,22 @@ export class Document implements IDocument {
...
+        this._patternPaths.forEach(path => {
+            if (path.isSelected()) {
+                this._sizeRatio = path.getLengthInPixels() / inputMeasurementInInches;
+                return;
+            }
+        });
+    }
+
+    getSizeRatio = (): number => {
+        if (this._sizeRatio === undefined) {
```
Comment (SarahPappas) [inline src/canvas/Document.tsx]: I think it is more common to use if (!this._sizeRatio)

File: src/canvas/Document.tsx [hunk 4]
```diff
@@ -28,4 +30,22 @@ export class Document implements IDocument {
         }
         return Boolean(this._patternPaths.pop());
     }
+
+    // Sets the pixels per inch ratio according to the input measurement
+    setSizeRatio = (inputMeasurementInInches: number): void => {
+        this._patternPaths.forEach(path => {
+            if (path.isSelected()) {
```
Comment (SarahPappas) [inline src/canvas/Document.tsx]: I think we should move selection to its own class, which holds a reference to the path that is selected. It doesn't make sense to put it on the path because it is temporary state. Additionally, it is more convenient to check which path is selected. Also, it is more flexible in case we want to dispatch events in the future when the selected path changes.

File: src/canvas/PatternPath.tsx [hunk 5]
```diff
@@ -9,6 +9,8 @@ export class PatternPath implements IPatternPath {
     private _path2D: Path2D;
     private _isPath2DValid: boolean;
     private _lastIndexAddedToPath2D: number;
+    private _isSelected: boolean;
```
Comment (SarahPappas) [inline src/canvas/PatternPath.tsx]: I think we should move isSelected and is Highlighted out to a different Selection class. I mentioned the reasons in my previous comment.

File: src/canvas/Renderer.tsx [hunk 6]
```diff
@@ -104,14 +140,20 @@ class Renderer implements IRenderer {
...
         context.lineJoin = 'round';
         context.lineCap = 'round';
         
         const paths = this._document.getPatternPaths();
         paths.forEach(path => {
             const path2D = path.getPath2D();
-            const pathColor = PatternPathColor.get(path.getType());
+            let pathColor: string | undefined;
+            if (path.isSelected() || path.isHighlighted()){
+                pathColor = '#000000';
```
Comment (SarahPappas) [inline src/canvas/Renderer.tsx]: We can add PatternPathColor.set(PatternPathType.Selected, '#0000000'); to PatternPathColor and use instead.

File: src/components/CanvasContainer/CanvasContainer.tsx [hunk 7]
```diff
@@ -9,13 +9,16 @@ interface CanvasContainerProps {
...
+  useEffect(() => {
     const canvas = renderer.init();
     if (canvasContainerRef.current) {
       canvasContainerRef.current.appendChild(canvas);
-      canvasContainerRef.current.style.backgroundImage = "url(" + uploadedFileData + ")";
     }
-  }, [canvasContainerRef, uploadedFileData]);
+  }, [canvasContainerRef]);
+
+  useEffect(() => {
```
Comment (SarahPappas) [inline src/components/CanvasContainer/CanvasContainer.tsx:22]: Why did you move this to another useEffect function?

File: src/pages/Trace/AddMeasurement/AddMeasurement.tsx [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+                // TODO: pop an error modal to the user
+                console.log('Please select a path');
+            } else {
+                console.log('ok');
+                pathDocument.setSizeRatio(parseFloat(inputMeasurement));
+            }
+        }
+    };
+
+    const instructModal: InstructModal = {text: ['Choose a path to measure.']};
```
Comment (SarahPappas) [inline src/pages/Trace/AddMeasurement/AddMeasurement.tsx]: Can we change this to 'Choose a line to measure.'

File: src/canvas/Renderer.tsx [hunk 9]
```diff
@@ -84,6 +84,42 @@ class Renderer implements IRenderer {
         return this._canvas;
     }
 
+    measurementInit = (): HTMLCanvasElement => {
```
Comment (SarahPappas) [inline src/canvas/Renderer.tsx]: measurementInit does not need to return an element, it can be void.

File: src/pages/Trace/AddMeasurement/AddMeasurement.tsx [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+export const AddMeasurement: React.FC<AddMeasurementProps> = ({ setUploadedFileData }) => {
+    const [inputMeasurement, setInputMeasurement] = React.useState<string>("");
+    
+    // Remove class that puts canvas in the background, remove picture,
+    // and reinitialize event listeners.
+    const canvasContainerRef = useRef(document.getElementsByClassName('canvasContainer')[0]);
+    useEffect(() => {
+        canvasContainerRef.current = document.getElementsByClassName('canvasContainer')[0];
+        canvasContainerRef.current.classList.remove('canvasContainerBackground');
+        canvasContainerRef.current.appendChild(renderer.measurementInit());
```
Comment (SarahPappas) [inline src/pages/Trace/AddMeasurement/AddMeasurement.tsx]: You don't need line 23. you can just call render.measurementInit(). (And measurementInit does not need to return an element)

File: src/pages/Trace/AddMeasurement/AddMeasurement.tsx [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+            } else {
+                console.log('ok');
+                pathDocument.setSizeRatio(parseFloat(inputMeasurement));
+            }
+        }
+    };
+
+    const instructModal: InstructModal = {text: ['Choose a path to measure.']};
+
+    const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
```
Comment (SarahPappas) [inline src/pages/Trace/AddMeasurement/AddMeasurement.tsx]: handleChange could be renamed  handleInput to be more specific.

File: src/canvas/Document.tsx (hunk 1 above)
Comment (msdrey) [reply src/canvas/Document.tsx]: ok

File: src/canvas/Document.tsx (hunk 2 above)
Comment (msdrey) [reply src/canvas/Document.tsx]: ok

File: src/canvas/Document.tsx (hunk 3 above)
Comment (msdrey) [reply src/canvas/Document.tsx]: ok

File: src/components/CanvasContainer/CanvasContainer.tsx (hunk 7 above)
Comment (msdrey) [reply src/components/CanvasContainer/CanvasContainer.tsx:22]: Because when we get to the measurement Page, we set uploadedFileData to "" to remove the photo from the brackground. This change triggered the useEffect that reinitialized the renderer uselessly and re-appended the canvas as a child of the canvasContainerRef unnecessarily, which made getting access to the canvas later weird and buggy

File: src/pages/Trace/AddMeasurement/AddMeasurement.tsx (hunk 8 above)
Comment (msdrey) [reply src/pages/Trace/AddMeasurement/AddMeasurement.tsx]: yes!

File: src/canvas/Renderer.tsx (hunk 9 above)
Comment (msdrey) [reply src/canvas/Renderer.tsx]: ah yes, I think that's a leftover of me trying things to solve some bugs. I will remove it.

File: src/pages/Trace/AddMeasurement/AddMeasurement.tsx (hunk 10 above)
Comment (msdrey) [reply src/pages/Trace/AddMeasurement/AddMeasurement.tsx]: ok

File: src/pages/Trace/AddMeasurement/AddMeasurement.tsx (hunk 11 above)
Comment (msdrey) [reply src/pages/Trace/AddMeasurement/AddMeasurement.tsx]: I disagree. Handlechange is triggered every time the user types a new character in the input, not on input submission. I would agree on something like handleInputChange though. Let me know what you think.

File: src/canvas/Renderer.tsx (hunk 6 above)
Comment (msdrey) [reply src/canvas/Renderer.tsx]: I thought about that that but it seemed weird since a PatternPathType is supposed to refer to weather a path is a seam, fold or edge, which has nothing to do with weather a path is selected or not.

Comment (msdrey) [reply src/canvas/Renderer.tsx]: I agree it shouldnt be set manually like it is now, though. I just havent found a way I like to do it.

File: src/canvas/Document.tsx (hunk 4 above)
Comment (msdrey) [reply src/canvas/Document.tsx]: I agree it's clunky and I did code it differently at first, but I wanted to keep as simple as possible in _drawPatternPaths in the renderer, because that gets called many times per second. how would your solution work there?

Comment (SarahPappas) [reply src/canvas/Document.tsx]: If we hold a reference in Selection, we could do something like
if(path === selection.selected)

File: src/components/CanvasContainer/CanvasContainer.tsx (hunk 7 above)
Comment (SarahPappas) [reply src/components/CanvasContainer/CanvasContainer.tsx:22]: I see, would you please add that explanation as a comment?

File: src/pages/Trace/AddMeasurement/AddMeasurement.tsx (hunk 11 above)
Comment (SarahPappas) [reply src/pages/Trace/AddMeasurement/AddMeasurement.tsx]: Ah, that's right. handleInputChange sounds good to me :)

File: src/canvas/Document.tsx (hunk 4 above)
Comment (msdrey) [reply src/canvas/Document.tsx]: ok, I tried something out and pushed it. I led to some weird stuff when accessing the selected path in addmeasurement, so I neeeded yet another getter in the renderer to access that. Let me know what you think.

File: src/pages/Trace/AddMeasurement/AddMeasurement.tsx (hunk 11 above)
Comment (msdrey) [reply src/pages/Trace/AddMeasurement/AddMeasurement.tsx]: ok, done

Comment (msdrey) [reply src/pages/Trace/AddMeasurement/AddMeasurement.tsx]: ok, done!

File: src/components/CanvasContainer/CanvasContainer.tsx (hunk 7 above)
Comment (msdrey) [reply src/components/CanvasContainer/CanvasContainer.tsx:22]: done!

File: src/canvas/Renderer.tsx (hunk 6 above)
Comment (msdrey) [reply src/canvas/Renderer.tsx]: We changed the map to accept PatterPathType or string as keys in order to specify a selected color without having to create a new PatternPathType.
